### PR TITLE
Add OAuth and error logging

### PR DESCRIPTION
**Pull request by knazart**
This adds more payment method logging for the OAuth process and for any errors on the health checks.

### Testing
* [ ] Connect (OAuth). There should be a log with the connection details (without the sensitive data, like tokens).
* [ ] Disconnect. Should see a log on disconnect.
* [ ] Refresh token process should also be logged.
* [ ] If any health check issues are encountered, it should be logged.

### CODE REVIEW

File: modules/EED_SquareOnsiteOAuth.module.php [hunk 1]
```diff
@@ -96,11 +96,26 @@ public static function requestAccess()
...
         // Check that we have all the required parameters and the nonce is ok.
         if (! wp_verify_nonce($_GET['nonce'], 'eea_square_grab_access_token')) {
-            // This is an error. Close the window.
-            EED_SquareOnsiteOAuth::closeOauthWindow(esc_html__('Nonce fail!', 'event_espresso'));
+            // This is an error. Log and close the window.
+            $err_msg = esc_html__('Nonce fail!', 'event_espresso');
+            EED_SquareOnsiteOAuth::errorLogAndExit($square_pm, $err_msg, [], true, true);
         }
+        // log the incoming data, don't exit
+        EED_SquareOnsiteOAuth::errorLogAndExit($square_pm, 'Request access request response', $_GET, false);
```
Comment (tn3rb) [inline modules/EED_SquareOnsiteOAuth.module.php:118]: do we need i18n for log messages? what have we done for other PMs?

```suggestion
        EED_SquareOnsiteOAuth::errorLogAndExit(
                $square_pm,
                esc_html__('Request access request response', 'event_espresso')
                $_GET, 
                false
        );
```

File: modules/EED_SquareOnsiteOAuth.module.php [hunk 2]
```diff
@@ -456,7 +460,7 @@ public static function refreshToken(EE_Payment_Method $squarePm)
                 } else {
                     $errMsg = esc_html__('Unknown response received!', 'event_espresso');
                 }
-                EED_SquareOnsiteOAuth::errorLogAndExit($squarePm, $errMsg, false);
+                EED_SquareOnsiteOAuth::errorLogAndExit($squarePm, $errMsg, (array) $responseBody, false);
```
Comment (tn3rb) [inline modules/EED_SquareOnsiteOAuth.module.php:463]: would be marginally better to typecast the `$responseBody` variable on line 446 where it is created instead of having to do so every time it is utilized.

also, not a huge deal, and something I likely missed because of all of the JS i've been dealing with recently, but PHP variables should be `camel_case` not `pascalCase`.  The one case where i think it kinda makes sense to break from that formatting schema is when you save a class to a variable, like `$EventModel`, but I don't even normally follow that.

File: modules/EED_SquareOnsiteOAuth.module.php [hunk 3]
```diff
@@ -556,12 +568,13 @@ public static function updateLocationsList(EE_Payment_Method $squarePm)
...
      */
-    public static function updateLocation(EE_Payment_Method $squarePm, array $locationsList)
+    public static function updateLocation(EE_Payment_Method $squarePm, array $locationsList): string
     {
         // Check if the selected location is saved, just in case merchant forgot to hit save.
         $location = $squarePm->get_extra_meta(Domain::META_KEY_LOCATION_ID, true);
         if (! $location || ! isset($locationsList[ $location ])) {
-            $defaultLocation = array_key_first($locationsList);
+            reset($locationsList);
+            $defaultLocation = key($locationsList);
```
Comment (tn3rb) [inline modules/EED_SquareOnsiteOAuth.module.php:577]: why the change?
Using `reset()`  will mutate the `$locationsList` (`$locations_list`) array, whereas `array_key_first()` does not cause side effects. Is that what you want?

File: modules/EED_SquareOnsiteOAuth.module.php [hunk 4]
```diff
@@ -721,37 +734,81 @@ public static function decryptString(string $text, bool $sandbox_mode): string
...
+            'nonce',
+        ];
+        foreach ($data as $name => $value) {
+            // for multi. arrays
+            if (is_array($value)) {
+                foreach ($value as $key => $val) {
+                    if (in_array($key, $sensitive_data)) {
+                        unset($data[ $key ]);
+                    }
+                }
```
Comment (tn3rb) [inline modules/EED_SquareOnsiteOAuth.module.php:757]: mmmm... i think you might have meant:

```suggestion
                foreach ($value as $key => $val) {
                    if (in_array($key, $sensitive_data)) {
                        unset($value[ $key ]);
                    }
                }
```

`$key` will surely not exist in the `$data` array since `$key` is coming from the inner array and `$data` is the parent array

File: modules/EED_SquareOnsiteOAuth.module.php [hunk 5]
```diff
@@ -721,37 +734,81 @@ public static function decryptString(string $text, bool $sandbox_mode): string
...
-     * @param EE_Payment_Method|boolean $squarePm
-     * @param null                      $errMsg
-     * @param bool                      $doExit Should we echo json and exit
-     * @param bool                      $alert Tells frontend to show an alert or not
+     * @param EE_Payment_Method|boolean $square_pm
+     * @param string                    $err_msg
+     * @param array                     $data
+     * @param bool                      $do_exit Should we echo json and exit
+     * @param bool                      $oauth
+     * @param bool                      $alert  Tells frontend to show an alert or not
```
Comment (tn3rb) [inline modules/EED_SquareOnsiteOAuth.module.php:775]: names for booleans that represent an action are more readable if the name reflects/infers the actions,
especially if they are completely self explanatory to the point where they don't require a comment 
(within reason of course, we don't want variable names that are tooooo long)

so maybe something like:

```suggestion
     * @param bool                      $echo_json_and_exit // or $echo_and_exit, or $echo_json, or $return_json
     * @param bool                      $using_oauth // or $oauth_enabled, or $oauth_active, or ???
     * @param bool                      $show_alert  Tells frontend to show an alert or not
```

File: modules/EED_SquareOnsiteOAuth.module.php [hunk 6]
```diff
@@ -721,37 +734,81 @@ public static function decryptString(string $text, bool $sandbox_mode): string
...
-        if ($squarePm instanceof EE_Payment_Method) {
-            $squarePm->type_obj()->get_gateway()->log(
-                ['Square error' => $errMsg],
+    public static function errorLogAndExit(
+        $square_pm,
+        string $err_msg = '',
+        array $data = [],
+        bool $do_exit = true,
+        bool $oauth = false,
+        bool $alert = false
```
Comment (tn3rb) [inline modules/EED_SquareOnsiteOAuth.module.php:785]: same point as above

File: modules/EED_SquareOnsiteOAuth.module.php [hunk 7]
```diff
@@ -777,4 +834,21 @@ public static function closeOauthWindow($msg = null)
...
+
+
+    /**
+     * Close the OAuth window with JS.
+     *
+     * @param string $err_msg
+     * @param bool   $alert
+     * @return void
+     */
+    public static function returnJsonError(string $err_msg = '', bool $alert = false)
```
Comment (tn3rb) [inline modules/EED_SquareOnsiteOAuth.module.php:846]: ACKCHYUALLY

technically this doesn't "return" the error, so maybe:

```suggestion
    public static function echoJsonError(string $err_msg = '', bool $alert = false)
```

**Review by tn3rb — COMMENTED** [review]
Found one potential issue that likely needs a fix
and also had a whole bunch of persnickety suggestions.

Otherwise looks great as always 👍🏻

File: modules/EED_SquareOnsiteOAuth.module.php (hunk 3 above)
Comment (knazart) [reply modules/EED_SquareOnsiteOAuth.module.php:577]: Simply trying to replace `array_key_first()` because that's `PHP7.3`.

Can also do an alternative:
```
foreach (locationsList as $location_name => $unused_val) {
    return $location_name;
}
```

File: modules/EED_SquareOnsiteOAuth.module.php (hunk 2 above)
Comment (knazart) [reply modules/EED_SquareOnsiteOAuth.module.php:463]: Oh yes, marginally better !

And yes, I know we should use `$var_name`, and I'm changing that in the code I'm working on, but I developed this plugin with our old PSR rules thus everything here is pascal case.

Comment (knazart) [reply modules/EED_SquareOnsiteOAuth.module.php:463]: Or maybe not that marginally better...
Square returns it as an object, so most methods deal with that response data as parameters of an object, thus changing all the code to use the data as an array just for the sake of this one or two methods that want an array might be a bad decision, as it would seem that the API response types are all over the place. That's just my opinion.

As an alternative, I could do the type casting inside the method itself.

File: modules/EED_SquareOnsiteOAuth.module.php (hunk 4 above)
Comment (knazart) [reply modules/EED_SquareOnsiteOAuth.module.php:757]: Oh, thanks for the `array_diff_key()` ! Didn't think of that one 🤦‍♂️

And yes, I was planning to use recursion and somehow dind't.

**Review by tn3rb — APPROVED** [review]
### looks great 🥇

**Comment by Pebblo** [comment]
I've been testing this and it seems to be working well so I'm mering it in before doing a full retest of Square.
